File: 1026_Maximum_Difference_Between_Node_and_Ancestor.java
// Leetcode 1026 - Maximum Difference Between Node and Ancestor
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

class Solution {
    public int maxAncestorDiff(TreeNode root) {
        // Start DFS with root value as both min and max
        return dfs(root, root.val, root.val);
    }

    // Helper function to traverse tree while tracking min and max ancestor values
    private int dfs(TreeNode node, int minv, int maxv) {
        if (node == null) {
            // When we reach leaf, return the max diff
            return maxv - minv;
        }

        // Update min and max values along the path
        minv = Math.min(minv, node.val);
        maxv = Math.max(maxv, node.val);

        // Recursively call for left and right subtrees
        int left = dfs(node.left, minv, maxv);
        int right = dfs(node.right, minv, maxv);

        // Return the greater of two
        return Math.max(left, right);
    }
}

File: 1026_Maximum_Difference_Between_Node_and_Ancestor.cpp
// Leetcode 1026 - Maximum Difference Between Node and Ancestor
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

class Solution {
    public:
        // Helper DFS function to traverse tree with current min and max values
        int dfs(TreeNode* node, int minv, int maxv) {
            if (!node) return maxv - minv;  // Return max difference at leaf
    
            // Update min and max with current node's value
            minv = min(minv, node->val);
            maxv = max(maxv, node->val);
    
            // Recursively compute for left and right subtrees
            int left = dfs(node->left, minv, maxv);
            int right = dfs(node->right, minv, maxv);
    
            // Return the max diff from both subtrees
            return max(left, right);
        }
    
        int maxAncestorDiff(TreeNode* root) {
            // Start DFS with root's value as initial min and max
            return dfs(root, root->val, root->val);
        }
    };
    